src/main/java/com/redhat/composer/services/ChatBotService.java

```diff
@@ -137,20 +164,63 @@ public Multi<String> chat(ChatBotRequest request) {
       });
       return multi;
     } catch (Exception e) {
-      log.error("Error in ChatBotService.chat", e);
+      Log.error("Error in ChatBotService.chat", e);
       return Multi.createFrom().failure(e);
     }
   }
 
   // TODO: Support non-streaming chat?
 
+  /**
+   * Prepare an AI service with zero or more content retrievers. Content retrievers may be from a
+   * {@link RetrieverRequest} or may be generated from provided {@link Document} that will be split, embedded,
+   * and stored in an {@link EmbeddingStore} for retrieval by the AI
+   *
+   * @param <T> the type of AI service to build
+   * @param aiServiceClass the class of AI service to build
+   * @param llm the language model to use in the AI service
+   * @param retrieverRequest an optional {@link RetrieverRequest} to provide additional knowledge to the AI
+   * @param documents an optional collection of {@link Document} in which the AI may infer knowledge
+   * @return the built AI service
+   */
+  private <T extends BaseAiService> T prepareAiService(
+      Class<T> aiServiceClass,
+      StreamingChatLanguageModel llm,
+      RetrieverRequest retrieverRequest,
+      Collection<Document> documents
+  ) {
+    AiServices<T> builder = AiServices.builder(aiServiceClass)
+        .streamingChatLanguageModel(llm);
+
+    List<ContentRetriever> retrievers = new ArrayList<>();
+
+    if (retrieverRequest != null) {
+      retrievers.add(
+          ragService.getContentRetriever(retrieverRequest)
+      );
+    }
+
+    if (documents != null && !documents.isEmpty()) {
+      retrievers.add(ragService.contentRetrieverForDocuments(
+          documents,
+          5,
```

Comment:
I am assuming this is temperature and like make return value? 

Could we make this configurable

Reply:
Yes, these are "magic numbers" which I'll fix. 5 is the maximum number of results to retrieve, and 0.75 is the minimum score. Where do you suggest they be configured? At the request level? At the retriever connection level? Or at the overall application level?

Reply:
I made it configurable at the application level, with some space to override application-level configuration if we decide to make it configurable per retriever connection.